### PR TITLE
Source position tracking

### PR DESCRIPTION
**Pull request by robdockins**
Retain source position information into the typechecked AST (for expressions) and use this information to print location information for runtime errors.

### CODE REVIEW

**Comment by robdockins** [comment]
CF #698, #984

**Comment by robdockins** [comment]
A couple of notes:

Currently, this PR preserves pretty much all of the location information present in the parsed expressions by directly translating the `ELocated` constructors.  We could probably get away with only annotating `EVar` constructors.

I attempted to implement call stacks in the same pass, but they turned out to be quite a bit more complicated.

File: src/Cryptol/Parser/Position.hs [hunk 1]
```diff
@@ -33,6 +33,11 @@ data Range      = Range { from   :: !Position
                         , source :: FilePath }
                   deriving (Eq, Ord, Show, Generic, NFData)
 
+type CallStack = [ (Text, Range) ]
```
Comment (robdockins) [inline src/Cryptol/Parser/Position.hs:36]: note to self: this stuff snuck in somehow

File: src/Cryptol/Backend/Monad.hs [hunk 2]
```diff
@@ -387,11 +399,23 @@ instance X.Exception Unsupported
 
 
 -- | For things like @`(inf)@ or @`(0-1)@.
-typeCannotBeDemoted :: Type -> a
-typeCannotBeDemoted t = X.throw (TypeCannotBeDemoted t)
+--typeCannotBeDemoted :: Type -> a
```
Comment (robdockins) [inline src/Cryptol/Backend/Monad.hs:402]: todo, remove these lines

File: src/Cryptol/Backend/Monad.hs [hunk 3]
```diff
@@ -262,7 +267,7 @@ unDelay tv =
                     -- a loop error.  If some other thread is evaluating, reset the
                     -- transaction to await completion of the thunk.
                     UnderEvaluation t _
-                      | tid == t  -> writeTVar tv (UnderEvaluation tid (X.throwIO (LoopError "")))
+                      | tid == t  -> writeTVar tv (UnderEvaluation tid (X.throwIO (EvalErrorEx emptyRange (LoopError "")))) -- TODO? better range info
```
Comment (robdockins) [inline src/Cryptol/Backend/Monad.hs:270]: probably should do something about this

**Comment by brianhuffman** [comment]
I just looked over the diff, and my main questions have to do with the new `Prim` type. At the very least, it could use some haddock documentation explaining what it's for. As far as I can tell, the main thing that distinguishes it from `GenValue` is the `PRange` constructor, so you can have values that depend on the source range they are evaluated in.

But it's not clear to me why this needs to be a separate type; is there a reason why we couldn't just add a `PRange`-style  constructor to `GenValue`?

**Comment by robdockins** [comment]
Couple of reasons.  First, the current source range is a _lexical_ concept, which is why it is passed around as an implicit parameter of `evalExpr` and `evalPrim`.  If something like `PRange` were a value form instead, the range would be provided _dynamically_ instead, and would depend on where the term was forced rather than where it was generated.  If we wanted this, we'd just add it to the `SEval` monad instead.  Second, it's a useful abstraction for some other things I want to do (CF #995, in particular) that mean I won't have to touch every primitive definition again when I want to change the representation of `VFun`, `VPoly` and `VFunPoly`.

File: src/Cryptol/Backend/Monad.hs [hunk 4]
```diff
@@ -163,27 +169,31 @@ maybeReady (Eval _) = pure Nothing
 --   its own evaluation.
 delayFill ::
   Eval a {- ^ Computation to delay -} ->
-  Eval a {- ^ Backup computation to run if a tight loop is detected -} ->
+  Maybe (Eval a) {- ^ Optional backup computation to run if a tight loop is detected -} ->
+  String {- ^ message for the <<loop>> exceprion if a tight loop is detecrted -} ->
+  Range {- ^ location information for the <<loop>> exceprion if a tight loop is detecrted -} ->
```
Comment (brianhuffman) [inline src/Cryptol/Backend/Monad.hs:174]: typo

**Comment by brianhuffman** [comment]
I think the `Backend` class could probably use some more haddock documentation explaining where the new `Range` arguments come from and what the class methods are supposed to do with them. It should explain enough so that someone who wanted to implement a new backend would know what to do.

**Review by brianhuffman — APPROVED** [review]
It could use a bit more documentation in a couple of spots, but otherwise looks good.

**Comment by robdockins** [comment]
Upon reflection, I think the state from #995 is a better ending point for these changes, now that I've figured out how to make call stacks work properly.

**Pull request closed without merging**